### ISSUE TITLE
Define high level API and tests

### ISSUE THREAD

**Issue by karalabe** [project-iris/iris-java#1]
The high level API should conform to http://godoc.org/github.com/project-iris/iris-go as much as possible.

Additionally, the high level tests should implement more or less https://github.com/project-iris/iris-go/tree/master/tests

**Comment by l0rinc** [comment]
Added most of the interfaces from the GO API to Java;

**Comment by l0rinc** [comment]
Added stubs for the higher layer, but should be finished TDD style;

**Comment by l0rinc** [comment]
Fixed the Broadcast and its test and added stubs for the rest of the functionality.
It's a pretty big mess though, it should be refactored once it's working (I don't understand everything so well to do it correctly from start, but once the tests cover the functionality, I will gladly play with it - got a few ideas already)

**Comment by karalabe** [comment]
We can grab a glass of wine when it's a bit more complete and work through it together :) I'm a tad busy with a conference review + Iris at dotScale + Iris at StrageLoop so I'll need to get the "deadlined" things sorted out first. Also it would be nice to get the second draft of the Iris relay protocol finished, so that is also a priority.

**Comment by karalabe** [comment]
Uuuuuu, teh drone.io tests are green for the Java binding. That's a first :D 
https://drone.io/github.com/project-iris/iris-java/20

**Comment by l0rinc** [comment]
Separated the callback interfaces so that you can define e.g. separate response processing for every request, and that users won't wonder why their tunnel callbacks aren't called for broadcasts :) (they were in the same global callback interface).

I've implemented and tested the request/response part also;
This means that a request/response with callback implementation may look like this:

``` Java
new Connection(IRIS_PORT, CLUSTER_NAME).request(CLUSTER_NAME, originalMessage, timeOutMillis, (requestId, receivedMessage) -> {
    System.out.println("Yey: " + Arrays.toString(receivedMessage));
});
```

I'm open to changes and suggestions (not claiming it's a final solution), but I think we're going in the right direction :)

**Comment by karalabe** [comment]
I think you misunderstood the design of the callbacks a bit. Think about two communicating nodes as client and server. The client issues the broadcasts/requests/tunnels while the server receives them and potentially responds.

Server side, you need a callback handler registered for every communication primitive, since you have no control over what the network will send you. Any type of message may arrive at any time. Hence why the callback handler is needed.

On the other hand, the client itself is always the one initiating the communication. In the case of broadcast and publish, it is a fire and forget mechanism. Now in the case of a request or a tunnel, the system will respond. But there is no reason for registering callbacks to these responses, as the original calls themselves should block until a response arrives or the operation times out.

Consider the [browser - webserver](http://play.iris.karalabe.com/talks/binds/go.slide#10) example from the demo presentation. The [webserver](http://play.iris.karalabe.com/talks/binds/go.slide#12) is our server here, which contains a callback registered for inbound requests. On the other hand, the [browser](http://play.iris.karalabe.com/talks/binds/go.slide#11) - out client - needs absolutely no callbacks, as the 'Request' will block until either a reply arrives or it times out. [Tunnels](http://play.iris.karalabe.com/talks/binds/go.slide#18) are analogous to this mode of operation (inbound tunnels pop up through callbacks, outbound tunnels are set up blockingly).

In your above design you turned the requests asynchronous client side too, but I think it is the wrong approach. Requests are meant to be very short and fast operations, mainly to fetch some data or do some short processing. As such, they are expected to run and return on the order of milliseconds. Calling code should wait for them to finish and then proceed with execution. A very simple example could be a database query: the client issues a request, but doesn't really know how to proceed until the response arrives. Just sit and wait will the reply arrives.

Now on the other hand there could be a notion of 'tasks' (i.e. long running operations) that would definitely benefit from such an asynchronous interface. But those would require significant support from the Iris runtime (i.e. decentralized, fault tolerant, etc that require significant work (among others, probably a Paxos implementation)). So that is off the chart right now.

To sum it up, I think you misunderstood and thought that replies also arrive through the connection's callback interface, which isn't how it's done. It simply gets returned as the result of the 'Request' function call :)

Hope it makes it clear why there shouldn't be callbacks when sending requests. If not, do ask :)

**Comment by karalabe** [comment]
PS: The request id must never bubble up to API level. It is an internal notion that the client of the lib never sees.

**Comment by karalabe** [comment]
PPS: Your example in the blocking logic:

``` java
System.out.println("Yey: " + Arrays.toString(new Connection(IRIS_PORT, CLUSTER_NAME).request(CLUSTER_NAME, originalMessage, timeOutMillis));
```

**Comment by l0rinc** [comment]
> But there is no reason for registering callbacks to these responses, as the original calls themselves should block until a response arrives or the operation times out
> [...]
> On the other hand, the browser - out client - needs absolutely no callbacks, as the 'Request' will block until either a reply arrives or it times out.

Isn't this the same thing? Both are callbacks, but one is blocking, the other is not. Am I missing something? You could get wrong response if you're not event driven (see: 'Init' if you have multiple Connection objects). I mean even if you send a request, the next _response_ could be for something else.
You can be event driven and blocking for some functions, but internally they can still be callbacks.

The way I am trying to design it currently is handlers for everything that can have a _response_ of some kind.
If there is no response to the client, the handler will automagically be deleted - no harm done.
If there is a response, it will be called asynchronously - or you can block until it arrives; same thing but unified.
The handlers can be throw-away handlers (i.e. get new ID every time, e.g. responses to requests) or static handlers (i.e. same ID, e.g. broadcast).

I would like to finish it this way, even if it ends up being rewritten, because the whole design seems a lot more uniform this way (in case I am solving the right problem, and not something else :p).

ps. these posts are editable :p

**Comment by karalabe** [comment]
I don't really see how you could get a wrong response with multiple connections. Each and every connection has it's own TCP link to the relay and is completely isolated from the rest. How could there be corruption between them?

The relay will _always_ respond to the client in the case of a request and tunnel. Either success, or a timeout, but a response will _always_ arrive. In the case of a broadcast and publish, replies are _impossible_.

Your design isn't inherently bad, you just designed for the rarer use case instead of the common one. Both blocking API can be turned async and vice versa, a non-blocking API turned sync. But the common use case is for requests to block, so why put the burden of blocking onto the client?

The thing you're trying to expose now is the multiplexing/demultiplexing logic (i.e. pushing all messages onto a single network link, and reading back the replies and matching them up with the original messages). This is something internal that should not concert the user of the library.
